subsys/net/lib/app/CMakeLists.txt

```diff
@@ -1,10 +1,10 @@
-zephyr_library()
```

Comment:
@SebastianBoe , I had to make this change as otherwise cmake complained (e.g. https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/17346/1/tests):

~~~
CMake Error at ../../../CMakeLists.txt:507 (message):
  The Zephyr library 'subsys__net__lib__app' was created without source
  files.  Empty (non-imported) libraries are not supported.  Either make sure
  that the library has the sources it should have, or make sure it is not
  created when it has no source files.
~~~

I'm not sure what are implications of converting this from "library" to just "a soup of files". Any suggestions?

Reply:
I would suggest that one determines under what conditions either NET_APP_SERVER or _CLIENT is enabled and then only add the net/lib/app subdirectory under this condition.

e.g.

if(CONFIG_NET_APP_SERVER OR CONFIG_NET_APP_CLIENT)
add_subdirectory(app)
endif()

Also, that

zephyr_link_interface_ifdef(CONFIG_MBEDTLS mbedTLS)

is removed. Each library that depends on mbedTLS should explicitly link with it because using it has "special" build requirements (it's headerfiles are not in zephyr/include).

Using the

zephyr_link_interface_ifdef(CONFIG_MBEDTLS mbedTLS)

API will unnecessarily corrupt the entire Zephyr build, instead of just the components that use mbedTLS.

Reply:
> I would suggest that one determines under what conditions either NET_APP_SERVER or _CLIENT is enabled and then only add the net/lib/app subdirectory under this condition.

I would find that rather unfortunate if cmake *requires* to do that. Because that obviously breaks an encapsulate of a subdirectory component: `app/` is tied to CONFIG_NET_APP, and it's just `app`'s internal detail that it further has CONFIG_NET_APP_SERVER and CONFIG_NET_APP_CLIENT. With your solution, parent CMakeLists.txt has to know about this internal detail, and they are thus duplicated in different places.

Overall, I appreciate the quick reply and suggestions, but there was no reply about the implications of the alternative solution of getting rid of a "cmake library". Nor it's clear why zephyr_link_interface_ifdef (which was there before, so I'd rather not touch it) would "corrupt" "the entire build". So, I can just blindly follow the suggestions without really learning much here, and the next time I'd have similar questions. (Yeah, I can do RTFM, but I of course tried to do that before, and it's not easy to find an answer to a specific question, and cmake's overall paradigm seems to be rather funky, leaving impression that it's in good deal a collection of workarounds, and suggestions like above only confirm that.)

Reply:
> With your solution, parent CMakeLists.txt has to know about this internal detail, and they are thus duplicated in different places.

To avoid this one could for instance use

add_subdirectory_ifdef(CONFIG_NET_APP app)

instead.

And then within the 'app' CMakeLists.txt

do

if(CONFIG_NET_APP_SERVER OR CONFIG_NET_APP_CLIENT)
zephyr_library()

[...]

endif()

Reply:
> Overall, I appreciate the quick reply and suggestions, but there was no reply about the implications of the alternative solution of getting rid of a "cmake library". Nor it's clear why  (which was there before, so I'd rather not touch it) would "corrupt" "the entire build". 

It's difficult to say what one should do, because the library is already doing something one shouldn't do. There is two options, do things the right way, or be safe and make a minimal change.

If you want to do things the right way.  Both invocations of zephyr_link_interface_ifdef should be removed. If you want to make the minimal possible change, then dropping the library is the best option.

The reason that zephyr_link_interface should not be used here is because it will cause the mbedtls include path to be added to all of the Zephyr sources, but it is only needed by a subset. The recommended way to solve this problem is to organize the code into Zephyr CMake libraries that link with each other and inherit build environment requirements from each other (like include paths).

Reply:
> The reason that zephyr_link_interface should not be used here is because it will cause the mbedtls include path to be added to all of the Zephyr sources, but it is only needed by a subset.

Thanks for the explanation. Well, now CI fails with (https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/17393/3/tests):

> /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/net/net_app.h:19:33: fatal error: config-mini-tls1_2.h: No such file or directory
 #include CONFIG_MBEDTLS_CFG_FILE

So, let me enable zephyr_link_interface_ifdef back and see if that what caused it.

Reply:
Hmm,

I see net_app.h includes mbedtls/config.h

That is unfortunate.

One public API (include/net/net_app.h) is including a non-public API.

net_app.h shouldn't be doing this. But since it is, it is correct to use zephyr_link_interface in this case. Because mbedTLS is a part of the public API when net_app is enabled.

Reply:
Thanks for confirming. So, if the build passes, I'll leave it in for now. We can look into it later (and net_app is heading for deprecation AFAIK, so we can just treat it as legacy with its own tricks now).

Reply:
To build nicely one should avoid components that are 

"Common X needed in various Y.

It is better to split X up and for each Y to include the specific parts of X that they need.
But until this is resolved, one can simply use zephyr_sources and zephyr_link_*.

Reply:
Ok, so just reenabling zephyr_link_interface_ifdef didn't help, there's the same CI error. I guess, it's indeed pretty convoluted, as CONFIG_NET_APP defaults to y, and then zephyr_link_interface_ifdef is on by default, and some apps assume they can build against mbedTLS "by default".

> But until this is resolved, one can simply use zephyr_sources and zephyr_link_*.

Yeah, until it's figured out (and mbedTLS integration as a whole needs various adjustments), let's just use which works now, and the above is effectively my initial solution, so I'm reverting back to it.